keep bluebird promise type-compatible with the lib.es6.d.ts promise

austinried opened the pull request:
This change keeps the Bluebird promise type-compatible with the ES6 promise definition found here: https://github.com/Microsoft/TypeScript/blob/master/lib/lib.es6.d.ts#L5505

I ran into this issue with WebStorm's version of those ES6 typings, but without the Symbol.toStringTag you get an error about the property being missing on Bluebird's promise interface when trying to mix the two.  Once it's added the issue is resolved.

lhecker commented:
@austinried Awesome solution! I have no idea why it solves my issues of converting a Bluebird Promise to a ES6 one, but it does work! Mostly...

I didn't work because typescript (version `2.1.0-dev.20160812`) complained about a non-matching type signature of `then()`. Can you replace [this](https://github.com/austinried/DefinitelyTyped/blob/270d8a0d348db966f5a651c78e0acd65ce97f9b6/bluebird/bluebird.d.ts#L50-L51) with the following and try again if it works with your setup?

```
  then<TResult1, TResult2>(onfulfilled: (value: T) => TResult1 | PromiseLike<TResult1>, onrejected: (reason: any) => TResult2 | PromiseLike<TResult2>): Promise<TResult1 | TResult2>;
  then<TResult>(onfulfilled: (value: T) => TResult | PromiseLike<TResult>, onrejected: (reason: any) => TResult | PromiseLike<TResult>): Promise<TResult>;
  then<TResult>(onfulfilled: (value: T) => TResult | PromiseLike<TResult>): Promise<TResult>;
  then(): Promise<T>;
```

I think (IMO) it'd also be a good idea to move the `[Symbol.toStringTag]: "Promise";` line way up as the first line of the class - even before the constructor, since it's basically an annotation for the type/kind of class. But that's your choice.

austinried commented:
@lhecker Hmm I actually get a lot of errors when I try to use those signatures from my `await` statements, which complain about the operand not having a valid callable 'then' member.  I converted those into the Bluebird types and came out with this:

```
  then<U, Q>(onfulfilled: (value: R) => U | Bluebird.Thenable<U>, onrejected: (reason: any) => Q | Bluebird.Thenable<Q>): Bluebird<U | Q>;
  then<U>(onfulfilled: (value: R) => U | Bluebird.Thenable<U>, onrejected: (reason: any) => U | Bluebird.Thenable<U>): Bluebird<U>;
  then<U>(onfulfilled: (value: R) => U | Bluebird.Thenable<U>): Bluebird<U>;
  then(): Bluebird<R>;
```

But I think the main issue there is that I'm not on TS 2.x yet, we're still using the stable release line of 1.8.x which seems to have different lib typings (lib.es2015.d.ts vs lib.es6.d.ts).  It's possible those may work on the TS version you're using though.  It seems like it'd be a good idea to update this `then` signature once TS 2.0 is released, and get just the Symbol.toStringTag in for now.  I've moved it to the top of the class as you suggested.

rbuckton commented:
@austinried Bluebird doesn't actually have a `Symbol.toStringTag` property. 

> without the Symbol.toStringTag you get an error about the property being missing on Bluebird's promise interface when trying to mix the two. 

How are you mixing the two? Are you trying to cast a Bluebird promise to an ES6 promise? Or are you trying to convert between the two?

If you are converting between the two, consider using `Promise.resolve` (available both in es6 and in Bluebird):

```ts
import BluebirdPromise = require("bluebird");
var bluebirdPromise: BluebirdPromise<boolean>;
var es6Promise: Promise<boolean>;

Promise.resolve(bluebirdPromise); // Promise<boolean>
BluebirdPromise.resolve(es6Promise); // BluebirdPromise<boolean>
```

austinried commented:
@rbuckton Trying to return a Bluebird promise as an ES6 promise, which used to work with Bluebird 2.0 and the typings for it.  I had thought that they were supposed to be compatible, which it seems they largely are, except it looks like you're right that Bluebird is missing the `Symbol.toStringTag` property.

Since that's the case I'll close this PR and settle for converting between the two.  Thanks for the help!
